src/cpu/x64/jit_uni_reorder.cpp

```diff
@@ -1281,12 +1579,200 @@ struct jit_uni_reorder_t : public primitive_t {
+                    dst_engine->kind(), dst_md);
+            if (_pd == nullptr) return status::out_of_memory;
+            if (_pd->init(engine, src_engine, dst_engine) != status::success) {
+                delete _pd;
+                return status::unimplemented;
+            }
+            _pd->prb_ = prb;
+            _pd->init_scratchpad_md();
+            return safe_ptr_assign<reorder_pd_t>(*reorder_pd, _pd);
+        }
+
+        tr::prb_t prb_;
+
+    private:
+        //
+        // Swap last two nodes, put block 4, 8, 16 nodes to first
+        //
+        static void prb_tile_normalize(tr::prb_t &p) {
+            if (!utils::one_of(p.nodes[0].n, 8ul, 16ul)
+                    && utils::one_of(p.nodes[1].n, 8ul, 16ul)) {
+                nstl::swap(p.nodes[0], p.nodes[1]);
+            }
+        }
+    };
+
+    jit_blk_reorder_t(const pd_t *apd) : primitive_t(apd) {
+        kernel_ = utils::make_unique<tr::jit_single_blk_kernel>(pd()->prb_);
+    }
+
+    size_t n(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return (int)pd()->prb_.nodes[d].n;
+    }
+    ptrdiff_t is(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return pd()->prb_.nodes[d].is;
+    }
+    ptrdiff_t os(int d) const {
+        assert(d < pd()->prb_.ndims);
+        return pd()->prb_.nodes[d].os;
+    }
+
+    status_t execute(const exec_ctx_t &ctx) const override {
+        auto in = CTX_IN_MEM(const char *, DNNL_ARG_FROM);
+        auto out = CTX_OUT_MEM(char *, DNNL_ARG_TO);
+
+        // kernel handle 2-dimension tiles, a tail is possible
+        auto &prb = this->pd()->prb_;
+        ptrdiff_t BH = 1;
+        for (int i = 2; i < prb.ndims; ++i) {
+            BH *= prb.nodes[i].n;
+        }
+
+        auto block_sz = n(0);
+        auto n1 = n(1);
+        auto i1 = is(1);
+        auto o1 = os(1);
+        auto FL = (n1 + block_sz - 1) / block_sz;
+        auto bh_stride = BH == 1 ? 0 : is(2);
+
+        auto itype_sz = data_type_size(pd()->prb_.itype);
+        auto otype_sz = data_type_size(pd()->prb_.otype);
+
+        const size_t work_amount = (size_t)BH * FL;
+        int nthr = adjust_num_threads(dnnl_get_max_threads(), work_amount);
+        parallel(nthr,
+                [BH, FL, block_sz, bh_stride, itype_sz, otype_sz, in, out, i1,
+                        o1, n1, this](int ithr, int nthr) {
+                    sse_guard g(alloca(sse_guard::stack_size));
+                    // Hail Marry on Win x64:
+                    // No xmm6 ~ xmm15 usage cross JIT function
+                    for_nd(ithr, nthr, BH, FL, [&](dim_t bh, dim_t fl) {
+                        auto fl_b = fl * block_sz;
+                        auto bh_b = bh_stride * bh;
+                        auto *i = in + (bh_b + fl_b * i1) * itype_sz;
+                        auto *o = out + (bh_b + fl_b * o1) * otype_sz;
+                        (*kernel_)(i, o, n1 - fl_b < block_sz);
+                    });
```

Comment:
I am not convinced this is a reliable way to work with calling convention.
Even though I agree it is highly unlikely, but still, the compiler is free to use xmm registers in between `sse_guard` and calling `kernel_` -- namely in implementation of `for_nd` functor.
In all honesty, I don't see any better options than putting `fl` loop under jit or just preserve these 10 registers on Windows (actually if I read the code correctly it is enough to preserve xmm6-xmm11, xmm15; xmm14 = xmm_tmp could be replaced with xmm0, so no need to preserve it).
Did you have a chance to see how that would affect performance?

Reply:
I considered it safe by a DUMPBIN operation on the binary and got no xmm6~15 usage inside this function, even in this compilation unit :).

if we put fl loop inside jit, it'll requires pre knowledge of threading. Which is not ideal since it may have different thread settings during primitive creation and execution.

Anyway, I'll try some performance comparison by conforming to the ABI first.

Reply:
> I considered it safe by a DUMPBIN operation on the binary and got no xmm6~15 usage inside this function, even in this compilation unit :).

Yeah, as I said, I hardly believe the compiler would do something crazy like this indeed. But still, we need to comply to the ABI unless we fully control the code. Which could be the case if we put fl into the jit; but this is hard, wouldn't be easy to maintain, etc... Hence, yeah, I would prefer to see if preserving the proper registers for Windows wouldn't affect performance much.